Test for correct restriction arguments in harvester

cmeessen opened the pull request:
I realised that a test for the correct spelling of the restriction directions (`north`, `south`, etc.) in `fatiando.harvester` is missing in #314.

Added _test_restriction() which tests for correct arguments. Test is coming soon.

### Checklist:

- [x] Make tests for new code (at least 80% coverage)
- [x] Create/update docstrings
- [x] Include relevant equations and citations in docstrings
- [x] Docstrings follow the style conventions
- [x] Code follows PEP8 style conventions
- [x] Code and docs have been spellchecked
- [x] Include new dependencies in `doc/install.rst`, `environment.yml`, `ci/requirements-conda.txt` and `ci/requirements-pip.txt`.
- [x] Documentation builds properly (run `cd doc; make` locally)
- [x] Changelog entry (leave for last to avoid conflicts)
- [x] Firt-time contributor? Add yourself to `doc/contributors.rst` (leave for last to avoid conflicts)

leouieda commented:
@cmeessen thanks for implementing this! 

I noticed that you're getting and error because `restrict` is not a list in your tests. I think the best and easiest way around this is to always pass restrict as a list (even if there is only one element). There is no need to enforce this on the code but maybe make it more clear in the docstring parameter description.

Let me know when you're done and I'll get this merged.

cmeessen commented:
@leouieda thank you for your comments. I changed the type of ``restrict`` to be a list by default and added the unit tests as well. ``_test_restriction()`` now also tests whether the argument is a list and raises a ``TypeError`` if not. All of this is also covered in the tests.

Not sure if these rather small changes should go into the changelog as well. If not, this PR is ready to merge.

leouieda commented on fatiando/gravmag/harvester.py:
```diff
@@ -221,7 +221,7 @@ def _find_index(point, mesh):
 
 def harvest(data, seeds, mesh, compactness, threshold, report=False,
-            restrict=None):
+            restrict=[None]):
```
@cmeessen it's really dangerous to use mutable python types (lists, dicts, etc)
as default function values. These are evaluated at import time and can then be
changed between successive calls to the function, for example:

    def func(y=[]):
        if y:
            return 'not empty'
        else:
            y.append('anything')
            return 'empty'


    >>> func()
    'empty'
    >>> func()
    'not empty'

That's why we always use `Ǹone` instead of an empty list or dict, even when the
required argument type is one of those.

I had a really hard to find bug in `fatiando.mesher` because I was setting
`props={}` as default. Values were changing unexpectedly and I had no idea why.

So you can leave this as `restrict=None`.

leouieda commented on fatiando/gravmag/harvester.py:
```diff
@@ -268,8 +268,8 @@ def harvest(data, seeds, mesh, compactness, threshold, report=False,
-        ``'west'``. You can pass in multiple directions as a list, e.g.
-        ``['above', 'north']``. Default is ``None`` for unrestricted growth.
+        ``'west'``. You can pass one or more directions as a list, e.g.
+        ``['above', 'north']``. Default is ``[None]`` for unrestricted growth.
```
It's best to exemplify here that even if there is only one argument it should be a list by including `['below']` as an example.

leouieda commented on fatiando/gravmag/harvester.py:
```diff
@@ -579,10 +579,13 @@ def _test_restriction(restrict):
+            if not(case in cases):
+                raise ValueError("Unrecognized item in restrict: %s" % case)
+    else:
+        raise TypeError("Argument restrict is not a list.")
```
Since `restrict=None` by default (see above), this check can be simpler:

    if restrict is None:
        return []
    else:
        for case in restrict:
            if case not in cases:
                raise ValueError(...)
    return restrict

Then there is no need to the type check and the code will work just as fine if `restrict` is some other iterable type, like a tuple or set.

In Python, it's generally best to avoid doing `type` checks. Instead, execute the code and it will break if you did something that isn't supported, like using an integer in a `for` loop.

Notice as well that the function now parses the `restrict` and returns the proper form (`[]` if it was `None`). Now you can put `restrict = _test_restrict(restrict)` in the main `harvester` code and remove the check if `restrict is None` from `_neighbor_indexes`.

cmeessen commented:
@leouieda thank you for the review and comments. When I read your comments I remembered you already told me that passing lists by default is dangerous.. I changed the code according to your suggestions. The approach returning an empty list by the function is quite neat.

cmeessen commented on fatiando/gravmag/harvester.py:
```diff
@@ -538,8 +539,6 @@ def _neighbor_indexes(n, mesh, restrict):
     nz, ny, nx = mesh.shape
     indexes = []
-    if restrict is None:
-        restrict = []
```
With _test_restriction() returning an empty list, this should be superfluous now.

leouieda commented:
@cmeessen perfect! Thanks so much for putting in the work! Sorry for the long review process.

My offer of free stickers is still valid. Just send me your address and I'll mail you some 🥇
